tox.ini

```diff
@@ -53,8 +53,12 @@ platform = {env:_PYTEST_TOX_PLATFORM:.*}
 [testenv:linting]
 skip_install = True
 basepython = python3
-deps = pre-commit>=1.11.0
-commands = pre-commit run --all-files --show-diff-on-failure
+deps =
+    pre-commit>=1.11.0
+    mypy==0.711
+commands =
+    pre-commit run --all-files --show-diff-on-failure
+    mypy bench/ extra/ scripts/ src/ testing/
```

Comment:
you can use mypy through pre-commit: https://github.com/pre-commit/mirrors-mypy

Reply:
I tried it initially, but couldn't get it to work. It errors with a lot of module name conflicts, like someone reported here: https://github.com/python/mypy/issues/4008.

It also says it runs in an isolated venv which means 3rd-party stubs can't be used.

Maybe it can be improved to fix these things?

Reply:
do you have specifics for the first? we use this at work and it works fine there
the second is solved using `additional_dependencies` -- no different from putting them in `deps` in `tox.ini`

Reply:
> do you have specifics for the first? we use this at work and it works fine there

Using this diff:

```diff
diff --git a/.pre-commit-config.yaml b/.pre-commit-config.yaml
index 0ff593289..4a53d7582 100644
--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -43,6 +43,10 @@ repos:
     rev: v1.4.0
     hooks:
     -   id: rst-backticks
+-   repo: https://github.com/pre-commit/mirrors-mypy
+    rev: v0.711
+    hooks:
+    -   id: mypy
 -   repo: local
     hooks:
     -   id: rst
```

mypy fails with errors like these:

```
doc/en/example/assertion/global_testmodule_config/conftest.py: error: Duplicate module named 'conftest' (also at 'testing/example_scripts/config/collect_pytest_prefix/conftest.py')
doc/en/example/assertion/global_testmodule_config/conftest.py: error: Are you missing an __init__.py?
```

Once you "fix" it (i.e. try to delete), another pops up.

>  the second is solved using additional_dependencies -- no different from putting them in deps in tox.ini

Right, it's the same. Though how easy would it be to get it to install the correct versions without repeating them?

Another thing is that it doesn't seem to pick up the configurations in `setup.cfg`. Can maybe use a `mypy.ini` and make it point to it? (I used a config file rather than flags so that code editors like vscode and manual `mypy` runs can pick them up automatically).

Reply:
> mypy fails with errors like these

ah yes that's PEP 420, if you add `__init__.py` files then mypy will recognize those as separate modules

> Right, it's the same. Though how easy would it be to get it to install the correct versions without repeating them?

you already have to repeat them in `tox` so it's no different

> Another thing is that it doesn't seem to pick up the configurations in `setup.cfg`

could you elaborate on "not respecting `setup.cfg`?" -- it's probably the same issue (`mypy-x.y.z` only works if there's `__init__.py` files in `x/`, `x/y`)

Reply:
the other thing you can do is split up the mypy invocations:

```yaml
    -   id: mypy
        name: mypy (src)
        files: ^src/
    -   id: mypy
        name: mypy (testing)
        files: ^testing/
    # ...
```

that way it doesn't try and dupe the modules